File: ios/Location/Sources/Location/GridReference.swift
//
//  GridReference.swift
//  Location
//
//  Created by Yanlin Li  on 5/9/2025.
//

// MARK: - GridReference

/// Represents the campus grid location of a building based on its ID.
public struct GridReference {
  public let campusCode: String
  public let sectionCode: String
  public let sectionNumber: Int
  public let campusSection: CampusSection

  /// Creates a grid reference by parsing a building ID.
  /// - Parameter buildingID: Building identifier in format "Campus-SectionNumber" (e.g., "K-J17")
  /// - Returns: GridReference with parsed campus and section information
  public static func fromBuildingID(buildingID: String) -> GridReference {
    let splitID = buildingID.split(separator: "-").map { String($0) }
    guard splitID.count == 2 else { return defaultGridReference() }

    let campusCode = splitID[0]
    let section = splitID[1]

    let startIndex = section.startIndex
    let secondIndex = section.index(after: startIndex)

    let sectionLetter = String(section[startIndex])
    guard let sectionNumber = Int(String(section[secondIndex...])) else {
      return defaultGridReference()
    }

    let campusSection = CampusSection(sectionNumber)

    return GridReference(
      campusCode: campusCode,
      sectionCode: sectionLetter,
      sectionNumber: sectionNumber,
      campusSection: campusSection)
  }

  public static func defaultGridReference() -> GridReference {
    GridReference(
      campusCode: "?",
      sectionCode: "?",
      sectionNumber: 0,
      campusSection: .upper)
  }
}

File: android/feature/buildings/src/main/java/com/devsoc/freerooms/feature/buildings/data/GridReference.kt
package com.devsoc.freerooms.feature.buildings.data

data class GridReference(
    val campusCode: String,
    val sectionCode: String,
    val sectionNumber: Int,
    val campusSection: CampusSection,
) {
    companion object {
        fun fromBuildingId(buildingId: String): GridReference {
            val parts = buildingId.split("-")
            if (parts.size != GRID_REFERENCE_PARTS) return default()

            val campusCode = parts[0]
            val section = parts[1]
            if (section.length < MIN_SECTION_LENGTH) return default()

            val sectionCode = section.first().toString()
            val sectionNumber = section.drop(1).toIntOrNull() ?: return default()

            return GridReference(
                campusCode = campusCode,
                sectionCode = sectionCode,
                sectionNumber = sectionNumber,
                campusSection = CampusSection.fromSectionNumber(sectionNumber),
            )
        }

        private fun default(): GridReference {
            return GridReference(
                campusCode = "?",
                sectionCode = "?",
                sectionNumber = 0,
                campusSection = CampusSection.UPPER,
            )
        }

        private const val GRID_REFERENCE_PARTS = 2
        private const val MIN_SECTION_LENGTH = 2
    }
}
